### PR TITLE
Codegen UPDATE

### PR DESCRIPTION
**Pull request by seojungmin**
This PR is for processing UPDATE within Codegen. The initial code has come from #656, and much of the code has been refactored and modified for a clean PR, e.g. the global runtime states introduced in the previous PR are all removed and localized, and a helper class like that in #687 has been introduced to accommodate many of the changes.  

`PerformUpdate()` that contains the original executor logic has been touched for a clear code representation.  Although it is more readable now, it is not still in its best condition.  There is also serialization/deserialization for the target/direct_map lists while being passed through the translator, and this may need some attention to be improved near future.

In addition, a bug on handling VARCHAR in codegen has been fixed, which is not related to UPDATE.   The original peloton value system has been including null terminators in its VARCHAR's length value, but that of codegen has been changed to behave in the same way.

### CODE REVIEW

**Comment by seojungmin** [comment]
The Coveralls program seemed to behave a little bit abnormally.  It reported that the coverage went down about 11%, which was unlikely considering the change that had been made in this PR.  @pmenon helped me to rerun the travis, and it seems to report the coverage correctly.

**Comment by seojungmin** [comment]
I will rebase this, now.

**Comment by seojungmin** [comment]
This PR can be merged. Review comments are always welcome.

**Comment by pmenon** [comment]
@seojungmin Can you provide high-level points on how this works? Also add comments in the code describing each component.

**Comment by seojungmin** [comment]
@pmenon Here are some high-level points. I added comments on the last commit.  

This PR introduces two new modules, and also changes roles of some previous modules:

[New]
- _Update translator_ builds the information required for update and hand it over to the updater.
- _Updater_ is a helper for the update execution, and deals with updating with and without primary key.  It also calls TransactionRuntime to find out(or to acquire) the ownership of the tuple. Depending on the ownership it has different execution strategies.  The execution strategy is the same as before, but the code is clean up.

[Role Changes]
- _TransactionRuntime_ used to do all the execution work, but I moved these roles to Deleter and I only left helper APIs for acquiring ownership.
- _Deleter_ used to be only a bridge to TransactionRuntime, but it now has all the execution logic inside, which is consistent with Updater.

Since he update execution logic was so huge and complex, these separations and changes were inevitable.   After all, the translators do whatever codegen can do, and pass over to Updater/Inserter/Deleter which actually do the execution jobs that cannot be handled by codegen.

File: src/codegen/operator/update_translator.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  auto &context = GetCompilationContext();
+  const auto *project_info = update_plan_.GetProjectInfo();
+  auto target_list = project_info->GetTargetList();
+  auto direct_map_list = project_info->GetDirectMapList();
+  uint32_t column_num =
+      static_cast<uint32_t>(target_list.size() + direct_map_list.size());
+  auto *scan =
+      static_cast<const planner::AbstractScan *>(update_plan_.GetChild(0));
+  std::vector<const planner::AttributeInfo *> ais;
+  scan->GetAttributes(ais);
```
Comment (pmenon) [inline src/codegen/operator/update_translator.cpp]: Shouldn't this be done in `UpdatePlan`? At this point, the plans should already know all attribute information.

File: src/codegen/updater.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+#include "storage/tile.h"
+#include "storage/tuple.h"
+#include "type/abstract_pool.h"
+#include "type/types.h"
+#include "type/value.h"
+
+namespace peloton {
+namespace codegen {
+
+void Updater::Init(concurrency::Transaction *txn, storage::DataTable *table,
```
Comment (pmenon) [inline src/codegen/updater.cpp]: Modify the initialization by removing `Transaction`, and add `ExecutorContext`. This way, you get both the memory pool and transaction handle at initialization time. You can then remove `GetPool()` and remove `ExecutorContext` from `Update()`

File: src/codegen/updater.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                   Target *target_vector, uint32_t target_vector_size) {
+  PL_ASSERT(txn != nullptr && table != nullptr && target_vector != nullptr);
+  txn_ = txn;
+  table_ = table;
+  // Target list is kept since it is required at a new version update
+  target_list_.reset(
+      new TargetList(target_vector, target_vector + target_vector_size));
```
Comment (pmenon) [inline src/codegen/updater.cpp]: Does `target_list_` have to be a unique_ptr?

File: src/include/codegen/operator/update_translator.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  // No state finalization
+  void TearDownState() override;
+
+  // Get the stringified name of this translator
+  std::string GetName() const override { return "Update"; }
+
+ private:
+  void SetTargetValue(llvm::Value *target_val_vec, llvm::Value *target_id,
+                      type::Type type, llvm::Value *value,
+                      llvm::Value *length) const;
```
Comment (pmenon) [inline src/include/codegen/operator/update_translator.h]: Not sure this is implemented.

File: src/include/codegen/operator/update_translator.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+ private:
+  // Plan
+  const planner::UpdatePlan &update_plan_;
+
+  // Target table
+  storage::DataTable *target_table_;
+
+  // Boolean value representing whether it is a primary key update or not
+  bool update_primary_key_;
```
Comment (pmenon) [inline src/include/codegen/operator/update_translator.h]: Not sure if this is used.

File: src/include/codegen/transaction_runtime.h [hunk 6]
```diff
@@ -13,6 +13,8 @@
 #pragma once
 
 #include <cstdint>
+#include "type/types.h"
+#include "common/item_pointer.h"
```
Comment (pmenon) [inline src/include/codegen/transaction_runtime.h]: Is this needed?

**Pull request closed without merging**